### ISSUE TITLE
`strict` option should take `env` into account

### ISSUE THREAD

**Issue by sindresorhus** [eslint/eslint#3306]
http://eslint.org/docs/rules/strict

## Problem

In Node.js it makes sense to use the `global` mode as files are implicitly wrapped in an IIFE by the runtime so you can safely put the use strict statement at the top of the file, but in the browser you'd want to use the `function` mode as you don't want to leak the use strict statement.

## Solution

Implement an `auto` mode to the `strict` option, that uses `global` if env is `node`, otherwise `function`.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
Rules cannot change their behavior based on which environments are used. Keep in mind, you can apply both `node` and `browser` environments at the same time. Can you explain more about what you're trying to do?

**Comment by silverwind** [comment]
The leaked strict mode is a concern when you bundle up scripts written for different modes.  I'm not sure if a `'use strict'` gets hoisted or not, but it sure will affect all scripts following it.

Unfortunately, it's probably impossible to determine whether a file is going to be ran in a browser or node.

**Comment by BYK** [comment]
I also have this problem and had to disable this rule when transitioning into v1.0.0

The problem is `Gruntfile` being run on NodeJS but most other files running in the browser via AMD. We want the "function" version for browser files and the global one for the ones running on NodeJS.

**Comment by sindresorhus** [comment]
@nzakas I want to be able to use a sharable config with the strict option and have the strict option adjust acordingly when I toggle an env on a file by file basis.

#### gulpfile (Node.js)

```js
/*eslint-env node */

...
```

#### main.js (browser)

```js
/*eslint-env browser */

...
```

**Comment by silverwind** [comment]
I think one way of solving this is to add an option like `"strict": [2, "env"]`, which then checks for an `eslint-env` comment in the file. If no env comment is given, there should be an error on line 0 to set one.

**Comment by nzakas** [comment]
As I mentioned, rules don't know anything about environments. That's by design, since rules shouldn't be keying off of environments, but rather key off of what it actually finds in the code or scope.

We might be able to do something using the `globalReturn` language feature, though. That's `true` for the node environment, so it's a soft indicator of the same thing. The question would still be how to provide an option that makes sense.

**Comment by btmills** [comment]
What about `smart`? Maps to `global` when `globalReturn` is `true`, otherwise `function`?

**Comment by sindresorhus** [comment]
> which then checks for an eslint-env comment in the file. If no comment is found, there should be an error on line 0 to set one.

The point isn't to enforce the `eslint-env` comment, but rather look at the resolved config for a file. If `env` contains `node` but not `browser`, enforce the "use strict" globally. If `env` contains `browser` but not `node`, enforce function scoped "use strict". If `env` contains both, do nothing.

**Comment by nzakas** [comment]
@btmills I kind of hate the term "smart" (one of the most overused for technical things: smart search, smart tabs, etc.), but I think you're thinking in the right direction. I'm also not sure I can come up with anything better, so I'm willing to go with it.

**Comment by btmills** [comment]
I agree with you there. Brainstorming a couple others: auto, detect, ...

**Comment by nzakas** [comment]
"safe"? As in, is safe to have " use strict" at the top level for Node but not for other places?

**Comment by btmills** [comment]
:+1: for "safe"
